File: 프로그래머스/lv2/60058. 괄호 변환/괄호 변환.py
def dfs(word):
    if word == "":
        return ""
    left = 0
    right = 0
    isCorrect = False
    if word[0] == "(":
        isCorrect = True
        left += 1
    else:
        isCorrect = False
        right += 1
    i = 1
    while left != right:
        if word[i] == "(":
            left += 1
        else:
            right += 1
        i += 1
    u = word[:i]
    v = word[i:]
    print("올바른 문자열 : ",isCorrect)
    print("u : ", u, "v : ", v)
    v = dfs(v)
    print("dfs 실행 결과 - u : ", u, "v : ", v)
    if isCorrect:
        answer = u + v
    else:
        answer = "(" + v + ")"
        for w in range(1, len(u)-1):
            if u[w] == "(":
                answer += ")"
            else:
                answer  += "("
    return answer

def solution(p):
    
    answer = dfs(p)
    return answer

File: 프로그래머스/lv2/60058. 괄호 변환/괄호 변환.cpp
#include <iostream>
#include <string>

using namespace std;

pair<string, string> splitUAndV(string word) {
    int left = 0;
    int right = 0;
    bool isCorrect = false;

    if (word[0] == '(') {
        isCorrect = true;
        left += 1;
    } else {
        isCorrect = false;
        right += 1;
    }

    int i = 1;
    while (left != right) {
        if (word[i] == '(') {
            left += 1;
        } else {
            right += 1;
        }
        i += 1;
    }

    string u = word.substr(0, i);
    string v = word.substr(i);

    return make_pair(u, v);
}

string dfs(string word) {
    if (word == "") {
        return "";
    }

    pair<string, string> uv = splitUAndV(word);
    string u = uv.first;
    string v = uv.second;

    v = dfs(v);

    if (u[0] == '(') {
        return u + v;
    } else {
        string answer = "(" + v + ")";
        for (int w = 1; w < u.length() - 1; w++) {
            if (u[w] == '(') {
                answer += ")";
            } else {
                answer += "(";
            }
        }
        return answer;
    }
}

string solution(string p) {
    string answer = dfs(p);
    return answer;
}